[lazy-defs] Add new reconstruction metadata API on Definitions

smackesey opened the pull request:
## Summary & Motivation

Add reconstruction metadata to `Definitions`, a new API for representing data underlying state-driven definitions from third-party sources. This allows potentially expensive external service calls underlying state-driven definitions to be made once in the code server, with the result reused when reconstructing `Definitions` again in child processes (run workers and step workers). Credit to @sryza for many of the ideas behind this API.

This is intended as a replacement for the existing `CacheableAssetsDefinition` API. Key differences:

- Code location metadata must be used with a `DefinitionsLoader` (`@definitions`) to work correctly
- Code location metadata can be used for metadata underlying any definition, not just assets.
- Code location metadata does not needs pseudo-definitions like `CacheableAssetDefinition` to ever be included in a `Definitions` object.

## Details

Consider this simple model of a third party service containing holding some data we wish to represent as assets in Dagster:

```python
WORKSPACE_ID = "my_workspace"

def fetch_foo_integration_asset_info(workspace_id: str):
    if workspace_id == WORKSPACE_ID:
        return [{"id": "alpha"}, {"id": "beta"}]
    else:
        raise Exception("Unknown workspace")
```

To represent it using source metadata:

```python
FOO_INTEGRATION_SOURCE_KEY = "foo_integration"

# This function would be provided by integration lib dagster-foo
def get_foo_integration_defs(context: DefinitionsLoadContext, workspace_id: str) -> Definitions:
    cache_key = f"{FOO_INTEGRATION_SOURCE_KEY}/{workspace_id}"
    if context.load_type == DefinitionsLoadType.RECONSTRUCTION and cache_key in context.reconstruction_metadata:
        payload = context.code_location_metadata[cache_key]
    else:
        payload = fetch_foo_integration_asset_info(workspace_id)
    asset_specs = [AssetSpec(item["id"]) for item in payload]
    assets = external_assets_from_specs(asset_specs)
    return Definitions(
        assets=assets,
    ).with_reconstruction_metadata(
        {cache_key: payload}
    )

@definitions
def defs(context: DefinitionsLoadContext):

    @asset
    def regular_asset(): ...

    return Definitions.merge(
        get_foo_integration_defs(context, WORKSPACE_ID),
        Definitions(assets=[regular_asset]),
    )
```

To represent this using the `CacheableAssetsDefinition` system:

```python
# This class would be provided by integration lib dagster-foo
class FooIntegrationCacheableAssetsDefinition(CacheableAssetsDefinition):
    def compute_cacheable_data(self) -> Sequence[AssetsDefinitionCacheableData]:
         payload = fetch_foo_integration_asset_info(self.unique_id)
         return [
            AssetsDefinitionCacheableData(extra_metadata={"id": item["id"]})
            for item in payload
        ]

    def build_definitions(self, data: Sequence[AssetsDefinitionCacheableData]):
        asset_specs = [AssetSpec(item.extra_metadata["id"]) for item in data]
        return external_assets_from_specs(asset_specs)

@definitions
def defs(context):

    @asset
    def regular_asset(): ...

    foo_integration_cacheable_assets = FooIntegrationCacheableAssetsDefinition(
        unique_id=WORKSPACE_ID
    )

    return Definitions(
        assets=[regular_asset, foo_integration_cacheable_assets],
    )
```

### Mechanism

- A `DefinitionsLoader` must be used for to define our `Definitions`. This is because only the loader function receives a `DefinitionsLoadContext`, which is the access point for reconstruction metadata.
- Integrations provide a function that accepts (a) a `DefinitionsLoadContext`; (b) credentials/API call params to retrieve a remote payload.
    - The function must return a `Definitions` object containing the third-party-sourced definitions.
    - In a reconstruction context, the function is expected to check `DefinitionsLoadContext.reconstruction_metadata` on the `DefinitionsLoadContext` for the payload and retrieve the remote payload on cache miss. Accesing `reconstruction_metadata` in an initialization context is an error.
    - The function adds metadata by calling `with_reconstruction_metadata()` on the returned object: `Definitions(...).with_metadata({<key>: <payload>})`. `with_reconstruction_metadata` wraps the passed value in `CodeLocationReconstructionMetadataValue` and sticks it on `Definitions.metadata`.
- If reconstruction metadata is stashed on a `Definitions`, then the `Definitions` object will generate a `PendingRepositoryDefinition` instead of a `RepositoryDefinition`. This is the same behavior as if any `CacheableAssetsDefinition` are detected in the definitions.
    - Using `PendingRepositoryDefinition` allows us to reuse the internal plumbing we already have for moving cached cached definition metadata cross-process: `RepositoryLoadData`.
- Any reconstruction metadata attached to `Definitions` gets attached to `RepositoryLoadData`, from which it is exposed on `DefinitionsLoadContext`. `RepositoryLoadData.cached_data_by_key` was renamed to `cacheable_asset_data`, since there are now two kinds of cached data. I didn't simply include source data in `cached_data_by_key` because the typing of this field is too restrictive, with values required to be `Sequence[AssetsDefinitionCacheableData]`.

## How I Tested These Changes

New unit tests.

## Changelog

New source metadata API for instantiating definitions backed by external APIs.

- [x] `NEW` _(added new feature or capability)_
- [ ] `BUGFIX` _(fixed a bug)_
- [ ] `DOCS` _(added or updated documentation)_

smackesey commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24395?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#24395** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24395?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#24337** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24337?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @smackesey and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

schrockn commented on python_modules/dagster/dagster/_core/definitions/repository_definition/repository_definition.py:
```diff
@@ -48,26 +48,36 @@
         "_RepositoryLoadData",
         [
-            ("cached_data_by_key", Mapping[str, Sequence[AssetsDefinitionCacheableData]]),
+            ("cacheable_asset_data", Mapping[str, Sequence[AssetsDefinitionCacheableData]]),
```
I don't see enough upside here to warrant a rename and dealing with storage back compat

smackesey replied:
Storage back compat is solved by the `storage_field_names` param above-- basically this is always serialized as/deserialized from the old name `cached_data_by_key`.

`cached_data_by_key` becomes very confusing if we have another field on here containing cached data.

schrockn replied:
We should not used "cached" in my view in the verbiage at all.

That all being said I don't care enough to block this PR on this issue.

smackesey replied:
>We should not used "cached" in my view in the verbiage at all.

The rename gets rid of "cached" so hopefully that's an improvement then. "cacheable" in the new name `cacheable_asset_data` is just to very explicitly link this data to its source in `CacheableAssetsDefinition`

schrockn reviewed (commented):
So looking at this a little more and thinking about it I do think we may want to provide a modicum more structure here. 

```python
def load_state(context, workspace_id):
   return fetch_foo_integration_asset_info(workspace_id)

def build_defs_from_state(state) -> Definitions: ...

def build_defs(context, workspace_id) -> Definitions
    state = load_defs_state(context, load_fn=load_state, state_key=workspace_id)
    return build_defs_from_state(state) # biz logic to build defs from state,
```

Or something like that. In particular I think the the framework should take responsibility for how to store the state, which will preserve a lot of option value for us.

schrockn reviewed (changes requested):
Per live discussion I think `cached_source_metadata` is not a great name for this. Also think even in the short-term we should consider a _slightly_ more structured API where state loading is done in a callback. We can take responsibility for storage that way.

schrockn commented on python_modules/dagster/dagster/_core/definitions/decorators/repository_decorator.py:
```diff
@@ -118,7 +118,9 @@ def __call__(
             bad_defns = []
             repository_defns = []
-            defer_repository_data = False
+            # If anv metadata is present, it is possible it contains cached info for defining
```
We need to make this much more explicit at a minimum. But I don't understand why we need this at all? In our structure, if you never use `CacheableAssetsDefinition` you shouldn't have to create a pending repository at all.

smackesey replied:
The short answer is that the simplest implementation for the goal of this PR made maximal use of the existing infrastructure, which runs through `PendingRepositoryDefinition`. The existing infrastructure uses `RepositoryLoadData` (which gets attached to `ExecutionPlanSnapshot`, which gets generated in the right place depending on whether there is a `PendingRepositoryDefinition`.

Getting away from `PendingRepositoryDefinition` will require additional complexity in this PR, which I would like to avoid for the MVP. `PendingRepositoryDefinition` is just an implementation detail here in any case.

schrockn replied:
Yeah ok. Well as per your other comment we definitely need to have a more precise semantic than this in the form of an explicit metadata value type or something.

smackesey replied:
Done, now works off detecting `CodeLocationStateMetadataValue`.

schrockn replied:
What about something more explicit that 

`RequiresPendingRepositoryLoadingMetadataValue` or something dumb like that.

schrockn commented on python_modules/dagster/dagster/_core/definitions/definitions_class.py:
```diff
@@ -730,3 +732,19 @@ def get_all_asset_specs(self) -> Sequence[AssetSpec]:
         asset_graph = self.get_asset_graph()
         return [asset_node.to_asset_spec() for asset_node in asset_graph.asset_nodes]
+
+    @experimental
+    def with_metadata(self, metadata: Mapping[str, Any]) -> Self:
+        """Add metadata to the Definitions object. This is typically used to cache data
+        loaded from some external API that is computed during execution of a DefinitionsLoader in a
+        code server. The cached data is then made available on the DefinitionsLoadContext when the
+        same target is loaded in another context (such as a run worker), allowing use of the cached
+        data to avoid additional external API queries.
+        """
+        return copy(
+            self,
+            metadata={
+                **(self.metadata or {}),
+                **metadata,
+            },
+        )
```
<!--__GRAPHITE_HTML_TAG_START__--><i class='graphite__hidden'>[Re: lines +736 to +751]<br/><br/></i>

<!--__GRAPHITE_HTML_TAG_END__-->shouldn't this be downstack?<!--__GRAPHITE_HTML_TAG_START__--><span class='graphite__hidden'><br/><br/>See this comment inline on <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24395?utm_source=unchanged-line-comment">Graphite</a>.</span>

<!--__GRAPHITE_HTML_TAG_END__-->

smackesey replied:
See my comments in that downstack PR. If we go with "metadata is private attribute only accessed through `with_metadata`", then yes, but I proposed something different.

schrockn replied:
roger

schrockn commented on python_modules/dagster/dagster/_core/definitions/definitions_loader.py:
```diff
@@ -26,6 +28,20 @@ class DefinitionsLoadContext:
 
+    @cached_property
+    def code_location_metadata(self) -> Mapping[str, Any]:
+        """Mapping[str, Any]: A dictionary containing metedata from the returned Definitions object
```
metadata

schrockn commented on python_modules/dagster/dagster/_core/definitions/definitions_loader.py:
```diff
@@ -26,6 +28,20 @@ class DefinitionsLoadContext:
     def __init__(self, repository_load_data: Optional["RepositoryLoadData"] = None):
         self._repository_load_data = repository_load_data
 
+    @cached_property
```
we should have an expliict state where this is available (when you are reconstructing an already-constructed code location), and it should throw if you try to access it when it is in an initial state where the code location doesn't exist yet. This is a bootstrapping problem.

smackesey replied:
Agree, this will require plumbing a "context type" or something into the `DefinitionsLoadContext` which can happen in another PR.

schrockn replied:
Define "next iteration". I think that context type or context state should be in an immediate follow up

smackesey replied:
Have now implemented in a downstack PR: https://github.com/dagster-io/dagster/pull/24490

schrockn reviewed (changes requested):
Very important to be expliict about the state of the context, and the lifecycle involved here.

schrockn reviewed (changes requested):
I want alignment on when we will be more explicit about tracking code location lifecycle state, which is an essential thing to get right immediately.

smackesey commented:
>I want alignment on when we will be more explicit about tracking code location lifecycle state, which is an essential thing to get right immediately.

OK I added a downstack PR that adds a `DefinitionsLoadType` that is accessible on the `DefinitionsLoadContext`.

Also implemented new `CodeLocationStateMetadataValue` class which is used to determine whether our `Definitions` has any "special" metadata.

After adjusting the APIs, I'm thinking that `reconstruction_metadata` (and `CodeLocationReconstructionMetadataValue`) is a more appropriate name than `code_location_metadata` (and `CodeLocationStateMetadataValue`). "Code location" alone isn't a great descriptor because we are trying to describe a specific subset of the metadata attached to a `Definitions` object.

schrockn commented on python_modules/dagster/dagster/_core/definitions/definitions_loader.py:
```diff
@@ -52,6 +53,24 @@ def load_type(self) -> DefinitionsLoadType:
+        at initial code location construction. This will be empty in the code server process, but in
+        child processes contain the metadata returned in the code server process.
+        """
+        if self._load_type != DefinitionsLoadType.RECONSTRUCTION:
```
excellent

schrockn commented on python_modules/dagster/dagster/_core/definitions/decorators/repository_decorator.py:
```diff
@@ -118,7 +119,13 @@ def __call__(
             bad_defns = []
             repository_defns = []
-            defer_repository_data = False
+            # If any CodeLocationStateMetadataValue is present, then it was set during code location
+            # construction and needs to be injected on reconstruction via RepositoryLoadData, so
+            # defer.
+            defer_repository_data = any(
+                isinstance(value, CodeLocationStateMetadataValue)
+                for value in self.metadata.values()
+            )
```
What is the line of sight to eliminating this? I am fine with this as a temporary but long term we should be able to just have this code path return vanilla `RepositoryDefinition`.

smackesey replied:
Just need to study the code more and find a convenient spot to construct `RepositoryLoadData` that doesn't rely on `PendingRepositoryDefinition`. Should be able to do it next week.

schrockn replied:
Ok. Can you create linear task for that and place a link to it in a comment.

schrockn commented on python_modules/dagster/dagster/_core/definitions/metadata/metadata_value.py:
```diff
@@ -1017,3 +1030,29 @@ class NullMetadataValue(NamedTuple("_NullMetadataValue", []), MetadataValue[None
+
+    def __new__(cls, data: Any):
+        try:
+            json.dumps(data)
```
hmmm shouldn't we be testing if it is packable by whitelist_for_serdes? Is json.dump a strict subset of that?

smackesey replied:
Pretty sure `json.dump` is a subset. JSON handles str, int, float, bool, list, dict, None. All of that is handled by serdes.

schrockn replied:
IMO we should just do an underlying `serdes` check since that is the true constraint.

schrockn commented on python_modules/dagster/dagster/_core/definitions/metadata/metadata_value.py:
```diff
@@ -1017,3 +1030,29 @@ class NullMetadataValue(NamedTuple("_NullMetadataValue", []), MetadataValue[None
+
+
+@whitelist_for_serdes
+class CodeLocationStateMetadataValue(
```
`ForDefinitionsReconstructionMetadataValue`? I think that might be best.

smackesey replied:
For the record we decided on `CodeLocationReconstructionMetadataValue`

schrockn commented on python_modules/dagster/dagster/_core/definitions/metadata/metadata_value.py:
```diff
@@ -1017,3 +1030,29 @@ class NullMetadataValue(NamedTuple("_NullMetadataValue", []), MetadataValue[None
+
+    @public
+    @property
+    def value(self) -> None:
```
fix type hint

schrockn reviewed (changes requested):
Let's do the new metadata value class name and some other comments

schrockn commented on python_modules/dagster/dagster/_core/definitions/decorators/definitions_decorator.py:
```diff
@@ -58,7 +58,7 @@ def fetch_foo_defs_metadata(workspace_id: str):
             def get_foo_defs(context: DefinitionsLoadContext, workspace_id: str) -> Definitions:
                 metadata_key = f"{FOO_METADATA_KEY_PREFIX}/{workspace_id}"
-                payload = context.code_location_metadata.get(metadata_key) or fetch_foo_defs_metadata(workspace_id)
+                payload = context.reconstruction_metadata.get(metadata_key) or fetch_foo_defs_metadata(workspace_id)
```
this example should be checking the load type of the context

schrockn commented on python_modules/dagster/dagster/_core/definitions/decorators/repository_decorator.py:
```diff
@@ -118,7 +121,13 @@ def __call__(
+            # If any CodeLocationStateMetadataValue is present, then it was set during code location
+            # construction and needs to be injected on reconstruction via RepositoryLoadData, so
+            # defer.
+            defer_repository_data = any(
+                isinstance(value, CodeLocationReconstructionMetadataValue)
+                for value in self.metadata.values()
+            )
```
please add link to linear task

schrockn reviewed (approved):
please heed final comments
